mbedtls/src/ssl/async_io.rs

```diff
@@ -95,17 +95,43 @@ where
     for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
 {
     fn poll_read(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>, buf: &mut ReadBuf<'_>) -> Poll<IoResult<()>> {
-        if self.handle().session.is_null() {
+        if self.handle().private_session.is_null() {
             return Poll::Ready(Err(IoError::new(IoErrorKind::Other, "stream has been shutdown")));
         }
 
-        self.with_bio_async(cx, |ssl_ctx| match ssl_ctx.recv(buf.initialize_unfilled()) {
-            Err(e) if e.high_level() == Some(codes::SslPeerCloseNotify) => Poll::Ready(Ok(())),
-            Err(e) if e.high_level() == Some(codes::SslWantRead) => Poll::Pending,
-            Err(e) => Poll::Ready(Err(crate::private::error_to_io_error(e))),
-            Ok(i) => {
-                buf.advance(i);
-                Poll::Ready(Ok(()))
+        self.with_bio_async(cx, |ssl_ctx| loop {
```

Comment:
Nitpick: I proposed a recursive call for the `read` function in `io.rs`. I'd take the same route here for consistency